sympy/functions/combinatorial/factorials.py

```diff
@@ -17,13 +17,13 @@
 class CombinatorialFunction(Function):
     """Base class for combinatorial functions. """
 
-    def _eval_simplify(self, **kwargs):
+    def _eval_simplify(self, measure=None , ratio=None):
```

Comment:
```suggestion
    def _eval_simplify(self, */, measure=None , ratio=None):
```

Reply:
It’s even better to use keyword only arguments such that users don’t pull wrong usage from ‘_eval_simplify(1, 2)’

Reply:
@sylee957 
I got your point but I am not getting why would there be any calls of the type  ‘_eval_simplify(1, 2)’ , since earlier it was **kwargs , which expects all arguments to be keyword . If at all there are any calls of the type ‘_eval_simplify(1, 2)’ , then even the existing code which tries to access like kwargs["measure"] would fail ? Please correct me if I am off :)

Reply:
I'd let you understand the theory about positional only arguments, keyword only arguments, and how these combines both
https://peps.python.org/pep-3102/

Simply changing `**kwargs` to `measure=None , ratio=None` makes `measure` and `ratio` both keyword argument as well as positional arguments.
I don't want to let users, or even usage in SymPy to do that, because that hurts reading and changing the codebase which is already overly complicated.

Reply:
I'd even suggest changing the arguments not have `None` default value in this case, 
because it doesn't make sense for multiplication

Reply:
@sylee957 
I got the point now . If I make the changes according to your suggestions would that be correct ....also how to interpret what shud be the default values  of the keyword args , since in this case there is no cue . I can say that default for measure can be None and for ratio it must be 1 .....since 1 is the multiplicative identity ?

Reply:
I may give `self, *, measure , ratio, **kwargs` as the best suggestion though.

One other point I have is that, `_eval_simplify` should be implemented like this:

```
def simplify(..., measure=..., ratio=..., **kwargs):
    ...
    _eval_simplify(measure=measure, ratio=ratio, **kwargs)
```

The reason that it 'carries' the keyword arguments is important. `measure, ratio` is part of recursive logic.
And if we use make some arguments optional `ratio=1`, 
it is highly likely that someone can ignore keywords  like `_eval_simplify()` by mistake 
and the implementation can have logical errors. (For example, if user explicitly passes `ratio=2`, then it should be passed exactly down to the functions)

If we just change the signature like above, we can get advantages that:

- It checks whether every keyword arguments `measure, ratio` are present when calling it. If it is not present, it will give errors as soon as possible.
- It checks whether every keyword arguments pass the value. It is less prone to mistakes because developers are asked to fill out `measure, ratio`, which gives them second chance to verify whether they are doing something strange for them. If the code just runs fine with optional arguments, it is more easy to be ignored and done mistakes for.

Reply:
@sylee957 @asmeurer 
But how  do I add the ```def simplify(..., measure=..., ratio=..., **kwargs):``` to factorials.py since it does not have a definition for it . Also once this change has been done would I need all calls to  _eval_simplify()  be replaced by simplify ().......?

Reply:
The point is to just make this change. 

```suggestion
    def _eval_simplify(self, *, measure=None , ratio=None, **kwargs):
```

Reply:
@asmeurer  thanks for the clarification .....Also I am facing following error with .mailmap file what can be done ?
```
Ambiguous names in .mailmap
This email address appears for multiple entries:
Person: PIAI23MAR4003 <patankarjays@gmail.com>
Mailmap entries:
Jay Patankar <patankarjays@gmail.com> Jay-Patankar <107458263+Jay-Patankar@users.noreply.github.com>

Lines should be added to .mailmap to indicate the correct name and
email aliases for all commits.

```

Reply:
@asmeurer @sylee957 
Also another doubt ........as I carefully read the commit logs certain tests are failing because there are additional keyword arguments like ```rational``` which may be passed through **kwargs , but I havent accounted for them . Your suggestions would solve that . But my question is while replacing any **kwargs with keyword arguments , how do I understand what all keyword arguments are supposed to be used , as there can be other arguments like ```rational```` which do not appear in the body and hence are difficult to track . An intelligent alternative would be to do something like what you have suggested here ```def _eval_simplify(self, *, measure=None , ratio=None, **kwargs):``` . But this does not make the signature free from **kwargs .

Reply:
That's expected. There could be arbitrary keyword arguments added in the future and we don't want to limit what the function can accept. But it should be explicit about the keyword arguments it does handle.

Reply:
https://github.com/sympy/sympy/pull/25752#discussion_r1342138481

I’d more likely to guide toward using keyword-only arguments without default than defaults, especially when the default values like None does not make sense for multiplication.

Reply:
If you don't include defaults, the values have to provided when the function is called.

Reply:
@asmeurer  @sylee957 
I got the point ....how about then keeping measure=None and ratio = 1 ..............that would make no harm for the multiplication I guess?

Reply:
@asmeurer can yu help me with https://github.com/sympy/sympy/pull/25752#discussion_r1343930394

Reply:
I have multiple arguments why defaults are fragile, and not good in this case.
And passing arguments comply with beijg more explicit.

https://github.com/sympy/sympy/pull/25752#discussion_r1342172119

Reply:
measure=None also does not make sense because None is not function.

I also don’t think that ratio should be 1. I think that simplify deliberately sets ratio to be 1.2 and has to be consistent with every callee

Reply:
@sylee957  okk got it .....still how do I get rid of the issue that I am facing with the .mailmap file?

Reply:
At the end of the day this method should only ever be called by simplify() and if simplify() always passes those arguments it doesn't really matter what the signature is Even `**kwargs` isn't that big of a deal. It's much more important for publicly facing methods.